File: 0101-0200/0127-Word Ladder/0127-Word Ladder.py
import collections
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        table = collections.defaultdict(list)
        L = len(beginWord)
        for word in wordList:
            for i in range(L):
                table[word[:i] + '*' + word[i + 1:]].append(word)
                
        Q = collections.deque([(beginWord, 1)])
        visited = set([beginWord])
        while Q:
            currw, step = Q.popleft()
            for i in range(L):
                nextw = currw[:i] + '*' + currw[i+1:]
                for word in table[nextw]:
                    if word == endWord:
                        return step + 1
                    if word not in visited:
                        Q.append((word, step + 1))
                        visited.add(word)
        return 0

# Bidirection
class Solution2:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        table = collections.defaultdict(list)
        L = len(beginWord)
        for word in wordList:
            for i in range(L):
                table[word[:i] + '*' + word[i + 1:]].append(word)
        
        qBegin = collections.deque([(beginWord, 1)])
        visitedBegin = {beginWord: 1}
        qEnd = collections.deque([(endWord, 1)])
        visitedEnd = {endWord: 1}
        def visitWord(Q, visited1, visited2):
            currw, step = Q.popleft()
            for i in range(L):
                nextw = currw[:i] + '*' + currw[i+1:]
                for word in table[nextw]:
                    if word in visited2:
                        return step + visited2[word]
                    if word not in visited1:
                        visited1[word] = step + 1
                        Q.append((word, step + 1))
                    
            return None
        while qBegin and qEnd:
            step = visitWord(qBegin, visitedBegin, visitedEnd)
            if step:
                return step
            step = visitWord(qEnd, visitedEnd, visitedBegin)
            if step:
                return step

        return 0


# Faster bidirection with swap
class Solution3:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        table = collections.defaultdict(list)
        L = len(beginWord)
        for word in wordList:
            for i in range(L):
                table[word[:i] + '*' + word[i + 1:]].append(word)
                
        step = 1
        visitedBegin = set([beginWord])
        visitedEnd = set([endWord])
        visited = set([beginWord, endWord])
        while visitedBegin and visitedEnd:
            if len(visitedBegin) > len(visitedEnd):
                visitedBegin, visitedEnd = visitedEnd, visitedBegin
            
            step += 1
            temp = set()
            for w in visitedBegin:
                for i in range(L):
                    nw = w[:i] + '*' + w[i+1:]
                    for word in table[nw]:
                        if word in visitedEnd:
                            return step
                        if word not in visited:
                            temp.add(word)
                            visited.add(word)                

            visitedBegin = temp

        return 0

File: 0101-0200/0127-Word Ladder/0127-Word Ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_map<string, vector<string>> table;
        int L = beginWord.size();
        for (string word : wordList) {
            for (int i = 0; i < L; ++i) {
                table[word.substr(0, i) + '*' + word.substr(i + 1)].push_back(word);
            }
        }
        
        int step = 1;
        queue<string> Q;
        Q.push(beginWord);
        unordered_set<string> visited{beginWord};
        while (!Q.empty()) {
            int qs = Q.size();
            ++step;
            for (int k = 0; k < qs; ++k) {
                string s = Q.front();
                Q.pop();
                
                for (int i = 0; i < L; ++i) {
                    string nextWord = s.substr(0, i) + '*' + s.substr(i + 1);
                    for (string word : table[nextWord]) {
                        if (word == endWord) {
                            return step;
                        }
                        if (!visited.count(word)) {
                            Q.push(word);
                            visited.insert(word);
                        }
                    }
                }
            }
        }
        
        return 0;
    }
};


class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        if (find(wordList.begin(), wordList.end(), endWord) == wordList.end()) {
            return 0;
        }
        unordered_map<string, vector<string>> table;
        int L = beginWord.size();
        for (string word : wordList) {
            for (int i = 0; i < L; ++i) {
                table[word.substr(0, i) + '*' + word.substr(i + 1)].push_back(word);
            }
        }
        
        int step = 1;
        unordered_set<string> visitedBegin{beginWord}, visitedEnd{endWord};
        unordered_set<string> visited{beginWord, endWord};
        while (!visitedBegin.empty() && !visitedEnd.empty()) {
            if (visitedBegin.size() > visitedEnd.size()) {
                swap(visitedBegin, visitedEnd);
            }
            
            ++step;
            unordered_set<string> temp;
            for (string w : visitedBegin) {
                for (int i = 0; i < L; ++i) {
                    string nw = w.substr(0, i) + '*' + w.substr(i + 1);
                    for (string word : table[nw]) {
                        if (visitedEnd.count(word)) {
                            return step;
                        }
                        if (!visited.count(word)) {
                            temp.insert(word);
                            visited.insert(word);
                        }
                    }
                }
            }
            visitedBegin = std::move(temp);
        }
        
        return 0;
    }
};
